src/Polysemy/Cont/Internal.hs

```diff
@@ -95,4 +103,155 @@ embedSem = liftSem . weave (pure ()) (pure . join) inspectSem
 {-# INLINE embedSem #-}
 
 newtype Ref m s a = Ref { runRef :: a -> m s }
+
+instance Contravariant (Ref m s) where
+  contramap f ref = Ref (runRef ref . f)
+
 newtype ExitRef m a = ExitRef { enterExit :: forall b. a -> m b }
+
+instance Contravariant (ExitRef m) where
+  contramap f ref = ExitRef $ \a -> enterExit ref (f a)
+
+data ViaFreshRef uniq a = ViaFreshRef { getBacktrackException :: a -> (uniq, Any) }
+
+instance Contravariant (ViaFreshRef uniq) where
+  contramap f ref = ViaFreshRef $ \a -> getBacktrackException ref (f a)
+
+{-
+  KingoftheHomeless: OK, so let's discuss how this works.
+  The idea here is to instead of providing a monadic computation
+  to the call of 'callCC' that simply short-circuits everything like
+  'ContT' does, we fake that behaviour by instead providing an
+  exception to 'callCC', and then try to 'catch' that exception
+  on the continuation. If the exception is caught, then we run the
+  continuation again. This way, we can get abortive continuations
+  without having to scope over a result type variable, avoiding
+  the problem that 'runContUnsafe' has, and making it possible
+  to weave other effects through without breaking everything.
+
+  Even with that solution, weaving effects through have more problems
+  of their own; namely, if we simply lower a
+  'forall s. ContT s (Sem r) a' to 'Sem r a', then we effectively
+  delimit all higher-order computations. This is bad, because
+  if a reified continuation produced within
+  the higher-order computation escapes from it,
+  then nothing can catch the underlying backtrack exception
+  once it is thrown.
+
+  The solution to this is anothor kludge: when weaving other effects through,
+  we instead use 'runContViaFreshInCWeave'; this makes use 'ContFreshState'
+  as its functorial state, which stores /handlers/ for backtrack exceptions.
+  'runContViaFreshInCWeave', in addition to 'catch'ing exceptions on the continuation
+  it is given, /also/ returns the handler it uses for the 'catch'.
+  This handler is then used by 'runContViaFresh' to catch exceptions on the
+  continuation /it/ gets, but can't provide to the higher-order computation.
+
+  I'm astonished that this even remotely works, but it does have some rather
+  weird behaviour I haven't completely figured out yet; especially in the
+  interaction with effects local in regards to the 'Cont'.
```

Comment:
This turns out to have been due to the issue https://github.com/polysemy-research/polysemy/pull/210 addresses. Still, the exact semantics of `runContViaFresh` still hasn't been figured out yet.